### PR TITLE
#211904 tags should take into account the FolderOrHost field or CT.ho…

### LINKED ISSUE #21904: Tags under specific site aren't listing
**Describe the bug**
If you create some tags under a specific site, that site Tag Storage is set to itself, when you create content and try to list those tags aren't showing.

**To Reproduce**
1. Create a blank site
2. Set the Tag Storage of that site to itself.
4. Create a Content-Type that lives under the new site, with text and a tag field.
5. Navigate to the Tags Portlet.
6. Add some tags that will live under the new site.
7. Create content and try to list the new tags that live under the specific site.

https://user-images.githubusercontent.com/5278970/160436470-6354419a-2112-48bd-a1ed-42d1e1633a2d.mov

**Workaround**
Add a site/folder field and select the specific site, that way the tags will be listed.

**Expected behavior**
I think should be like this:
* If the Content-Type lives under the specific site and there isn't a site/folder field, we should be able to list tags under it and also global ones.
* If the Content-Type lives under the System Host and there isn't a site/folder field, we should be able to list only global tags.
* If the Content-Type lives under the specific site or System Host and there is a site/folder field, we should be able to list tags under the site that is selected in the field and the global ones.

**Additional context**
Customer ticket: https://dotcms.zendesk.com/agent/tickets/106884
Reproducible in 21.06.7 LTS and 22.03.

### PR DESCRIPTION
**Pull request by fabrizzio-dotCMS**
The logic should be if the CT has a field of type site or folder then we use the site from that box. if not we should use the host field on the content type we're editing

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/rest/TagResource.java [hunk 1]
```diff
@@ -61,59 +107,428 @@ public Map<String, RestTag> list(@Context final HttpServletRequest request,@Cont
...
+        final String userId = tagForm.getOwnerId();
+        final ImmutableList.Builder<Tag> savedTags = ImmutableList.builder();
+        final ImmutableList.Builder<ErrorEntity> saveFails = ImmutableList.builder();
+        final Map<String, RestTag> tags = tagForm.getTags();
+
+        for (Entry<String, RestTag> entry : tags.entrySet()) {
+            final String tagKey = entry.getKey();
+            final RestTag tag = entry.getValue();
+            final String siteId = helper.getValidateSite(tag.siteId,user, request);
+            try {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rest/TagResource.java:155]: add some log

File: dotCMS/src/main/java/com/dotcms/rest/TagResource.java [hunk 2]
```diff
@@ -61,59 +107,428 @@ public Map<String, RestTag> list(@Context final HttpServletRequest request,@Cont
...
+                        .requiredBackendUser(true)
+                        .requiredFrontendUser(false)
+                        .requestAndResponse(request, response)
+                        .rejectWhenNoUser(true)
+                        .requiredPortlet(TAGS)
+                        .init();
+
+        final User user = initDataObject.getUser();
+
+        tagForm.checkValid();
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rest/TagResource.java:198]: this is usually called automatically by the framework

File: dotCMS/src/main/java/com/dotcms/rest/TagResource.java [hunk 3]
```diff
@@ -61,59 +107,428 @@ public Map<String, RestTag> list(@Context final HttpServletRequest request,@Cont
...
+        final Tag tagByTagId = Try.of(() -> tagAPI.getTagByTagId(tagForm.tagId)).getOrNull();
+        if (null == tagByTagId) {
+            final String errorMessage = Try.of(() -> LanguageUtil
+                    .get(user.getLocale(), "tag.id.not.found", tagForm.tagId))
+                    .getOrElse(String.format("Tag with id %s wasn't found.",
+                            tagForm.tagId)); //fallback message
+            throw new DoesNotExistException(errorMessage);
+        }
+        try {
+            tagAPI.updateTag(tagForm.tagId, tagForm.tagName, false, tagForm.siteId);
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rest/TagResource.java:220]: add log

File: dotCMS/src/main/java/com/dotcms/rest/TagResource.java [hunk 4]
```diff
@@ -61,59 +107,428 @@ public Map<String, RestTag> list(@Context final HttpServletRequest request,@Cont
...
+                new WebResource.InitBuilder(webResource)
+                        .requiredBackendUser(true)
+                        .requiredFrontendUser(false)
+                        .requestAndResponse(request, response)
+                        .rejectWhenNoUser(true)
+                        .requiredPortlet(TAGS)
+                        .init();
+
+        final User user = initDataObject.getUser();
+
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rest/TagResource.java:263]: Log here

File: dotCMS/src/main/java/com/dotcms/rest/TagResource.java [hunk 5]
```diff
@@ -61,59 +107,428 @@ public Map<String, RestTag> list(@Context final HttpServletRequest request,@Cont
...
+                new WebResource.InitBuilder(webResource)
+                        .requiredBackendUser(true)
+                        .requiredFrontendUser(false)
+                        .requestAndResponse(request, response)
+                        .rejectWhenNoUser(true)
+                        .requiredPortlet(TAGS)
+                        .init();
 
-        return siteId;
+        final ImmutableList.Builder<Tag> builder = ImmutableList.builder();
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rest/TagResource.java:300]: Logs

File: dotCMS/src/main/java/com/dotcms/rest/TagResource.java [hunk 6]
```diff
@@ -61,59 +107,428 @@ public Map<String, RestTag> list(@Context final HttpServletRequest request,@Cont
...
+                    String.format("update tag `%s` data is invalid or incomplete ", tagForm.tagId));
+            final String errorMessage = Try
+                    .of(() -> LanguageUtil.get(user.getLocale(), "tag.save.error.default", tagForm.tagId))
+                    .getOrElse(String.format("update tag `%s` data is invalid or incomplete .",
+                            tagForm.tagId)); //fallback message
+            throw new BadRequestException(errorMessage);
+        }
+
+        //We can assign tags to any user as long as we are admin.
+        final Tag tagByTagId = Try.of(() -> tagAPI.getTagByTagId(tagForm.tagId)).getOrNull();
```
[the remark this answers is not in the document]
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotcms/rest/TagResource.java:211]: trure. but I wanted to return null and add build a custom message and then log it and then trow the exception which is also perfectly valid

**Review by nollymar — APPROVED** [review]

**Review by jcastro-dotcms — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/rest/TagResource.java (hunk 3 above)
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotcms/rest/TagResource.java:222]: done

File: dotCMS/src/main/java/com/dotcms/rest/TagResource.java (hunk 4 above)
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotcms/rest/TagResource.java]: done

**Pull request merged**